src/Microsoft.AspNetCore.Blazor.Server/Circuits/CircuitSynchronizationContext.cs

```diff
@@ -0,0 +1,269 @@
+            // in this context. Let's instead add it to the queue and yield.
+            //
+            // We use our own queue here to maintain the execution order of the callbacks scheduled here. Also
+            // we need a queue rather than just scheduling an item in the thread pool - those items would immediately
+            // block and hurt scalability.
+            //
+            // We need to capture the execution context so we can restore it later. This code is similar to
+            // the call path of ThreadPool.QueueUserWorkItem and System.Threading.QueueUserWorkItemCallback.
+            ExecutionContext executionContext = null;
+            if (!ExecutionContext.IsFlowSuppressed())
+            {
+                executionContext = ExecutionContext.Capture();
+            }
+
+            return antecedant.ContinueWith(BackgroundWorkThunk, new WorkItem()
+            {
+                SynchronizationContext = this,
+                ExecutionContext = executionContext,
+                Callback = d,
+                State = state,
+            }, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Current);
+        }
+
+        private void ExecuteSynchronously(
+            TaskCompletionSource<object> completion,
+            SendOrPostCallback d,
+            object state)
+        {
+            var original = Current;
+            try
+            {
+                SetSynchronizationContext(this);
+                _state.IsBusy = true;
+
+                d(state);
+            }
+            finally
+            {
+                _state.IsBusy = false;
+                SetSynchronizationContext(original);
+
+                completion?.SetResult(null);
+            }
+        }
+
+        private void ExecuteBackground(WorkItem item)
+        {
+            if (item.ExecutionContext == null)
+            {
+                try
+                {
+                    ExecuteSynchronously(null, item.Callback, item.State);
+                }
+                catch (Exception ex)
+                {
+                    DispatchException(ex);
+                }
+
+                return;
+            }
+
+            // Perf - using a static thunk here to avoid a delegate allocation.
+            try
+            {
+                ExecutionContext.Run(item.ExecutionContext, ExecutionContextThunk, item);
+            }
+            catch (Exception ex)
+            {
+                DispatchException(ex);
+            }
+        }
+
+        private void DispatchException(Exception ex)
+        {
+            var handler = UnhandledException;
+            if (handler != null)
+            {
+                handler(this, new UnhandledExceptionEventArgs(ex, isTerminating: false));
+            }
+        }
+
+        private class State
+        {
+            public bool IsBusy; // Just for debugging
+            public object Lock = new object();
+            public Task Task = Task.CompletedTask;
```

Comment:
I played with using ValueTask here and it's not incredibly possible due to the fact that we want to avoid allocations when we execute synchronously - since the value task is returned at the end of the method. 

We'd basically end up inventing task completion source since it's needs to be a heap object - due to send being sync you need to be able to communicate with multiple pending work items.

I think this is the best we'll be able to do for now if we want to be non-blocking.

Reply:
Sure. I wouldn't worry about this massively! In practice we'll be enqueuing work items when we receive inbound interop calls, which will then typically perform a render, serialize the result, and send an outbound notification, all of which will do more allocations (hopefully not loads, but some). One further `Task` allocation here isn't likely to be measurable.